executor/analyze_test.go

```diff
@@ -1233,3 +1233,256 @@ func TestAnalyzeSamplingWorkPanic(t *testing.T) {
+	require.Equal(t, len(rs.Rows()), 1)
+	require.Equal(t, "0.9", rs.Rows()[0][2])
+	require.Equal(t, "3", rs.Rows()[0][3])
+	require.Equal(t, "1", rs.Rows()[0][4])
+	require.Equal(t, "LIST", rs.Rows()[0][5])
+	colIDStrs := strings.Join([]string{strconv.FormatInt(tableInfo.Columns[0].ID, 10), strconv.FormatInt(tableInfo.Columns[1].ID, 10)}, ",")
+	require.Equal(t, colIDStrs, rs.Rows()[0][6])
+
+	// disable option persistence
+	tk.MustExec("set global tidb_persist_analyze_options = false")
+	// manual analyze will neither use the pre-persisted options nor persist new options
+	tk.MustExec("analyze table t with 2 topn")
+	tbl = h.GetTableStats(tableInfo)
+	require.Greater(t, tbl.Version, lastVersion)
+	col0 = tbl.Columns[tableInfo.Columns[0].ID]
+	require.NotEqual(t, 3, len(col0.Buckets))
+	// The columns are: table_id, sample_num, sample_rate, buckets, topn, column_choice, column_ids.
+	rs = tk.MustQuery("select * from mysql.analyze_options where table_id=" + strconv.FormatInt(tbl.PhysicalID, 10))
+	require.Equal(t, len(rs.Rows()), 1)
+	require.NotEqual(t, "2", rs.Rows()[0][4])
+}
+
+func TestSavedPartitionAnalyzeOptions(t *testing.T) {
+	store, clean := testkit.CreateMockStore(t)
+	defer clean()
+
+	tk := testkit.NewTestKit(t, store)
+	tk.MustExec("use test")
+	tk.MustExec("set @@session.tidb_analyze_version = 2")
+	tk.MustExec("set global tidb_persist_analyze_options = true")
+	createTable := `CREATE TABLE t (a int, b int, c varchar(10), primary key(a), index idx(b))
+PARTITION BY RANGE ( a ) (
+		PARTITION p0 VALUES LESS THAN (8),
+		PARTITION p1 VALUES LESS THAN (20)
+)`
+	tk.MustExec(createTable)
+	tk.MustExec("insert into t values (1,1,1),(2,1,2),(3,1,3),(4,1,4),(5,1,5),(6,1,6),(7,7,7),(8,8,8),(9,9,9)")
+
+	h := domain.GetDomain(tk.Session().(sessionctx.Context)).StatsHandle()
+	oriLease := h.Lease()
+	h.SetLease(1)
+	defer func() {
+		h.SetLease(oriLease)
+	}()
+	tk.MustExec("analyze table t partition p0 with 1 topn, 2 buckets")
+	is := tk.Session().(sessionctx.Context).GetInfoSchema().(infoschema.InfoSchema)
+	tk.MustQuery("select * from t where b > 1 and c > 1")
+	require.NoError(t, h.LoadNeededHistograms())
+	table, err := is.TableByName(model.NewCIStr("test"), model.NewCIStr("t"))
+	require.Nil(t, err)
+	tableInfo := table.Meta()
+	pi := tableInfo.GetPartitionInfo()
+	require.NotNil(t, pi)
+	p0 := h.GetPartitionStats(tableInfo, pi.Definitions[0].ID)
+	lastVersion := p0.Version
+	col0 := p0.Columns[tableInfo.Columns[0].ID]
+	require.Equal(t, 2, len(col0.Buckets))
+	// The columns are: table_id, sample_num, sample_rate, buckets, topn, column_choice, column_ids.
+	rs := tk.MustQuery("select * from mysql.analyze_options where table_id=" + strconv.FormatInt(p0.PhysicalID, 10))
+	require.Equal(t, len(rs.Rows()), 1)
+	require.Equal(t, "2", rs.Rows()[0][3])
+	require.Equal(t, "1", rs.Rows()[0][4])
+	rs = tk.MustQuery("select * from mysql.analyze_options where table_id=" + strconv.FormatInt(tableInfo.ID, 10))
```

Comment:
`analyze table t partition p0 with 1 topn, 2 buckets` will update not only partition `p0`'s analyze options but also table `t`'s?

Reply:
For example, maybe the data distribution of `p0` is special, so we `analyze table t partition p0 with 1 topn, 2 buckets` to set its special options. Table `t` also records the special options. Then we execute `analyze table t`. After that, the special options `with 1 topn, 2 buckets` spread from table level to each partition. Is seems kind of strange.

Reply:
Had a discussion with Eason about it too. Then we decide that analyze table t partition p0 with optionlist will only update p0. table t still keeps the old options, also uses the old options to generate global stats. And our stats merging logic support this way.

The basic logic of this design is that, in most of the cases, people do 2 steps, 1. analyze table t with optionlist for most of the partitions 2. and then analyze one or two specific partitions with different optionlist since the special data distribution of the partitions.